### PR TITLE
Create global config file for storing credentials

### PR DESCRIPTION
**Pull request by alexcb**
The global config file (default location ~/earthly/config.yaml) can be
used to store https credentials for various sites. For example:

```yaml
git:
  default:
    auth: https
    username: foo
    password: bar
  github.com:
    username: f00
    password: bart
```

If the --git-username or --git-password command line values are given,
then those values will override all values stored in the config file.

### CODE REVIEW

**Comment by alexcb** [comment]
`example of git clone over ssh failing (due to me not having a ssh key configured)

alex@mah:~/gh/earthly/earthly$ ./build/linux/amd64/earth --config /dev/null gitlab.com/alexcb/privproj+build
loading config values from "/dev/null"
buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)
buildkitd | Settings do not match. Restarting buildkit daemon with updated settings...
buildkitd | ...Done
git clone git@gitlab.com:alexcb/privproj.git
ERROR: failed to fetch remote git@gitlab.com:alexcb/privproj.git: exit status 128
ERROR: failed to authorize: failed to fetch anonymous token: context canceled: context canceled
Error: resolve build context for target gitlab.com/alexcb/privproj+build: failed to solve: rpc error: code = Unknown desc = failed to load cache key: failed to fetch remote git@gitlab.com:alexcb/privproj.git: exit status 128
Check your git auth settings.
Did you ssh-add today? Need to specify GIT_USERNAME and GIT_PASSWORD?
For more information see https://docs.earthly.dev/guides/auth


example of it working over https with command line args

alex@mah:~/gh/earthly/earthly$ ./build/linux/amd64/earth --config /dev/null --git-username alexcb --git-password secret gitlab.com/alexcb/privproj+build
loading config values from "/dev/null"
buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)
buildkitd | Settings do not match. Restarting buildkit daemon with updated settings...
buildkitd | ...Done
git clone git@gitlab.com:alexcb/privproj.git
context | --> git context gitlab.com/alexcb/privproj+build
context | --> cache context gitlab.com/alexcb/privproj:master
gitlab.com/alexcb/privproj:master+base | --> FROM ubuntu:20.10
gitlab.com/alexcb/privproj:master+base | resolve docker.io/library/ubuntu:20.10@sha256:f43b0fe840eb118a74bdde75cd01def7344a5a2e628084ff30873d274d181721 0%
gitlab.com/alexcb/privproj:master+base | resolve docker.io/library/ubuntu:20.10@sha256:f43b0fe840eb118a74bdde75cd01def7344a5a2e628084ff30873d274d181721 100%
gitlab.com/alexcb/privproj:master+base | *cached* --> RUN [apt-get update && apt-get install -y build-essential cmake]
gitlab.com/alexcb/privproj:master+base | *cached* --> WORKDIR /code
gitlab.com/alexcb/privproj:master+code | *cached* --> COPY [src] src
gitlab.com/alexcb/privproj:master+build | *cached* --> RUN [cmake src]
gitlab.com/alexcb/privproj:master+build | *cached* --> RUN [make]
gitlab.com/alexcb/privproj:master+build | Target gitlab.com/alexcb/privproj:master+build built successfully
=========================== SUCCESS ===========================

example working with config file over https with config file

alex@mah:~/gh/earthly/earthly$ cat test.yaml 
git:
    gitlab.com:
        auth: https
        user: alexcb
        password: "secret"
    github.com:
        auth: https
        user: alexcb
        password: "password2"

alex@mah:~/gh/earthly/earthly$ ./build/linux/amd64/earth --config test.yaml gitlab.com/alexcb/privproj+build
loading config values from "test.yaml"
buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)
buildkitd | Settings do not match. Restarting buildkit daemon with updated settings...
buildkitd | ...Done
git clone git@gitlab.com:alexcb/privproj.git
context | --> git context gitlab.com/alexcb/privproj+build
context | --> cache context gitlab.com/alexcb/privproj:master
gitlab.com/alexcb/privproj:master+base | --> FROM ubuntu:20.10
gitlab.com/alexcb/privproj:master+base | resolve docker.io/library/ubuntu:20.10@sha256:f43b0fe840eb118a74bdde75cd01def7344a5a2e628084ff30873d274d181721 0%
gitlab.com/alexcb/privproj:master+base | resolve docker.io/library/ubuntu:20.10@sha256:f43b0fe840eb118a74bdde75cd01def7344a5a2e628084ff30873d274d181721 100%
gitlab.com/alexcb/privproj:master+base | *cached* --> RUN [apt-get update && apt-get install -y build-essential cmake]
gitlab.com/alexcb/privproj:master+base | *cached* --> WORKDIR /code
gitlab.com/alexcb/privproj:master+code | *cached* --> COPY [src] src
gitlab.com/alexcb/privproj:master+build | *cached* --> RUN [cmake src]
gitlab.com/alexcb/privproj:master+build | *cached* --> RUN [make]
gitlab.com/alexcb/privproj:master+build | Target gitlab.com/alexcb/privproj:master+build built successfully
=========================== SUCCESS ===========================
`

File: build.earth [hunk 1]
```diff
@@ -33,6 +33,7 @@ code:
 	COPY --dir buildcontext builder cleanup cmd conslogging dockertar domain \
 		llbutil logging ./
 	COPY --dir buildkitd/buildkitd.go buildkitd/settings.go buildkitd/
+	COPY --dir config/*.go config/
```
Comment (vladaionescu) [inline build.earth:36]:
```suggestion
	COPY --dir buildcontext builder cleanup cmd config conslogging dockertar domain \
		llbutil logging ./
	COPY --dir buildkitd/buildkitd.go buildkitd/settings.go buildkitd/
```

File: buildcontext/git.go [hunk 2]
```diff
@@ -89,7 +89,7 @@ func (gr *gitResolver) resolveGitProject(ctx context.Context, target domain.Targ
 		return nil, "", "", fmt.Errorf("Invalid github project path %s", target.ProjectPath)
 	}
 	githubUsername := projectPathParts[0]
-	githubProject := projectPathParts[1]
+	githubProject := strings.TrimSuffix(projectPathParts[1], ".git")
```
Comment (vladaionescu) [inline buildcontext/git.go:92]: Hmm.. this should not be necessary? `.git` is not allowed in target references.

File: cmd/earth/main.go [hunk 3]
```diff
@@ -14,11 +15,13 @@ import (
 	"github.com/earthly/earthly/builder"
 	"github.com/earthly/earthly/buildkitd"
 	"github.com/earthly/earthly/cleanup"
+	"github.com/earthly/earthly/config"
 	"github.com/earthly/earthly/conslogging"
 	"github.com/earthly/earthly/domain"
 	"github.com/earthly/earthly/earthfile2llb"
 	"github.com/earthly/earthly/earthfile2llb/variables"
 	"github.com/earthly/earthly/logging"
+
```
Comment (vladaionescu) [inline cmd/earth/main.go:24]: Nit: Unnecessary blank line?

File: buildkitd/entrypoint.sh [hunk 4]
```diff
@@ -34,6 +34,23 @@ if [ -n "$GIT_URL_INSTEAD_OF" ]; then
...
 
+# setup git credentials and config
+i=0
+while true
+do
+    varname=GIT_CREDENTIALS_"$i"
+    eval data=\$$varname
+    if [ -n "$data" ]
+    then
+        echo $data | base64 -d > /usr/bin/git_credentials_"$i"
```
Comment (vladaionescu) [inline buildkitd/entrypoint.sh:45]: Hmm.. we shouldn't write these to disk. Can we just keep them as env vars and never write them as files in the clear?

File: cmd/earth/main.go [hunk 5]
```diff
@@ -39,20 +42,23 @@ type earthApp struct {
...
+	pruneAll            bool
+	pruneReset          bool
+	buildkitdSettings   buildkitd.Settings
+	allowPrivileged     bool
+	buildkitHost        string
+	buildkitdImage      string
+	remoteCache         string
+	configPath          string
+	GitUsernameOverride string
+	GitPasswordOverride string
```
Comment (vladaionescu) [inline cmd/earth/main.go:61]: I believe these can be private, right?

File: cmd/earth/main.go [hunk 6]
```diff
@@ -163,6 +166,13 @@ func newEarthApp(ctx context.Context, console conslogging.ConsoleLogger) *earthA
 			Usage:       "Do not use cache while building",
 			Destination: &app.noCache,
 		},
+		&cli.StringFlag{
+			Name:        "config",
+			Value:       "~/earthly/config.yaml",
```
Comment (vladaionescu) [inline cmd/earth/main.go:171]: Should `~` be replaced by `os.Getenv("HOME")`?

File: cmd/earth/main.go (hunk 3 above)
Comment (alexcb) [reply cmd/earth/main.go:24]: I intentionally added this line as a way to separate earthly imports from thirdparty imports, but we can revert back to having them all together. I have no strong feeling either way.

File: cmd/earth/main.go (hunk 6 above)
Comment (alexcb) [reply cmd/earth/main.go:171]: good catch!

File: cmd/earth/main.go (hunk 5 above)
Comment (alexcb) [reply cmd/earth/main.go]: definitely, not sure why I made them private. fixed.

File: buildkitd/entrypoint.sh (hunk 4 above)
Comment (alexcb) [reply buildkitd/entrypoint.sh]: I changed the helper to echo out (and decode) it from the env now.

File: buildcontext/git.go (hunk 2 above)
Comment (alexcb) [reply buildcontext/git.go]: I accidentally tried running this:
```
$ earth github.com/alexcb/hello.git+docker
buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)
ERROR: failed to fetch remote git@github.com:alexcb/hello.git.git: exit status 128
```
and got confused. I wonder if we should error out sooner saying the target is not valid.

File: buildkitd/entrypoint.sh [hunk 7]
```diff
@@ -34,6 +34,23 @@ if [ -n "$GIT_URL_INSTEAD_OF" ]; then
...
+    if [ -n "$data" ]
+    then
+        echo echo \$$varname \| base64 -d > /usr/bin/git_credentials_"$i"
+        chmod +x /usr/bin/git_credentials_"$i"
+    else
+        break
+    fi
+    i=$((i+1))
+done
+echo $GIT_CONFIG | base64 -d > /root/.gitconfig
```
Comment (vladaionescu) [inline buildkitd/entrypoint.sh:52]:
```suggestion
echo "$GIT_CONFIG" | base64 -d > /root/.gitconfig
```

Nit - It doesn't really matter here, more like a style comment.

File: cmd/earth/main.go [hunk 8]
```diff
@@ -262,9 +272,64 @@ func newEarthApp(ctx context.Context, console conslogging.ConsoleLogger) *earthA
...
+func (app *earthApp) parseConfigFile(context *cli.Context) error {
+	if context.IsSet("config") {
+		app.console.Printf("loading config values from %q\n", app.configPath)
+	}
+
+	yamlData, err := ioutil.ReadFile(app.configPath)
+	if os.IsNotExist(err) && !context.IsSet("config") {
+		yamlData = []byte{}
+	} else if err != nil {
+		return err
```
Comment (vladaionescu) [inline cmd/earth/main.go:289]: Would be good to `errors.Wrapf` here to say which file we've tried to read.

File: cmd/earth/main.go [hunk 9]
```diff
@@ -262,9 +272,64 @@ func newEarthApp(ctx context.Context, console conslogging.ConsoleLogger) *earthA
...
+	}
+
+	cfg, err := config.ParseConfigFile(yamlData)
+	if err != nil {
+		return err
```
Comment (vladaionescu) [inline cmd/earth/main.go:294]: Similar here - which file we tried to parse

File: cmd/earth/main.go [hunk 10]
```diff
@@ -262,9 +272,64 @@ func newEarthApp(ctx context.Context, console conslogging.ConsoleLogger) *earthA
...
+			if app.gitPasswordOverride != "" {
+				v.Password = app.gitPasswordOverride
+			}
+			cfg.Git[k] = v
+		}
+	}
+
+	gitConfig, gitCredentials, err := config.CreateGitConfig(cfg)
+	if err != nil {
+		return err
```
Comment (vladaionescu) [inline cmd/earth/main.go:324]: Similar here

File: config/config.go [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+	return &config, nil
+}
+
+func CreateGitConfig(config *Config) (string, []string, error) {
+	credentials := []string{}
+	lines := []string{}
+	cred_i := 0
+
+	// iterate over map in a consistent order otherwise it will cause the buildkitd image to restart
+	// due to the settings hash being different
```
Comment (vladaionescu) [inline config/config.go:55]: Good thinking ;-)

**Pull request merged**